src/main/java/com/commercetools/sync/inventories/InventorySync.java

```diff
@@ -0,0 +1,444 @@
+                    statistics.incrementFailed();
+                    handleFailure("Failed to process inventory entry without sku", null);
+                }
+            } else {
+                statistics.incrementProcessed();
+                statistics.incrementFailed();
+                handleFailure("Failed to process null object", null);
+            }
+        }
+        if (!accumulator.isEmpty()) {
+            completableFutures.add(processBatch(accumulator).toCompletableFuture());
+        }
+        return CompletableFuture
+            .allOf(completableFutures.toArray(new CompletableFuture[completableFutures.size()]))
+            .thenApply(v -> statistics);
+    }
+
+    @Override
+    @Nonnull
+    public InventorySyncStatistics getStatistics() {
+        return statistics;
+    }
+
+    /**
+     * Methods tries to
+     * <ul>
+     *     <li>Fetch existing supply channels from CTP project</li>
+     *     <li>Instantiate {@code supplyChannelKeyToId} map</li>
+     *     <li>Create missing supply channels if needed</li>
+     * </ul>
+     * Method returns {@link CompletionStage} of {@link Void} that indicates method progress. It may contain exception
+     * occurred during fetching supply channels.
+     *
+     * @param drafts {@link List} containing {@link InventoryEntryDraft} objects where missing supply channels can occur
+     * @return {@link CompletionStage} of {@link Void} that indicates method progress. It may contain exception
+     *     occurred during fetching supply channels
+     */
+    private CompletionStage<Void> populateSupplyChannels(@Nonnull final List<InventoryEntryDraft> drafts) {
+        return inventoryService.fetchAllSupplyChannels()
+                .thenAccept(existingSupplyChannels ->
+                    supplyChannelKeyToId = existingSupplyChannels.stream()
+                            .collect(toMap(Channel::getKey, Channel::getId))
+                )
+                .thenCompose(v -> createMissingSupplyChannels(drafts));
+    }
+
+    /**
+     * When {@code ensureChannel} from {@link InventorySyncOptions} is set to {@code true} then attempts to create
+     * missing supply channels. Missing supply channel is a supply channel of key that can not be found in CTP project,
+     * but occurs in {@code drafts} list. Method returns {@link CompletionStage} of {@link Void} that indicates all
+     * possible creation attempts progress.
+     *
+     * @param drafts {@link List} containing {@link InventoryEntryDraft} objects where missing supply channels can occur
+     * @return {@link CompletionStage} of {@link Void} that indicates all possible creation attempts progress
+     */
+    @SuppressFBWarnings("NP_NONNULL_PARAM_VIOLATION") // https://github.com/findbugsproject/findbugs/issues/79
+    private CompletionStage<Void> createMissingSupplyChannels(@Nonnull final List<InventoryEntryDraft> drafts) {
+        if (syncOptions.isEnsureChannels()) {
```

Comment:
In `InventorySyncOptionsTest` I would rather have options created explicit by builder instead of using `getInventorySyncOptions` shortcut, because the test concerns that builder. In fact only one test case there could actually use thismethod. `getInventorySyncOptions` is used heavily by `InventorySyncTest` and there is no need to move it for now IMO.